HW2/P5/physics_sp2.pyx

```diff
@@ -0,0 +1,176 @@
+                            for dim in range(2):
+                                V2[dim] += eps * (XY2[dim] - XY1[dim])
+
+cpdef update(FLOAT[:, ::1] XY,
+             FLOAT[:, ::1] V,
+             UINT[:, ::1] Grid,
+             float R,
+             float grid_spacing,
+             uintptr_t locks_ptr,
+             float t,
+             int nthreads):
+    cdef:
+        int count = XY.shape[0]
+        int i, j, dim, new_grid_x, new_grid_y, old_grid_x, old_grid_y
+        int chunk_size = count / nthreads 
+        FLOAT *XY1, *XY2, *V1, *V2
+        # SUBPROBLEM 4: uncomment this code.
+        # omp_lock_t *locks = <omp_lock_t *> <void *> locks_ptr
+
+    assert XY.shape[0] == V.shape[0]
+    assert XY.shape[1] == V.shape[1] == 2
+
+    with nogil:
+        # bounce off of walls
+        #
+        # SUBPROBLEM 1: parallelize this loop over 4 threads, with static
+        # scheduling.
+        for i in prange(count, num_threads=nthreads, chunksize=chunk_size, schedule='static'):
+            for dim in range(2):
+                if (((XY[i, dim] < R) and (V[i, dim] < 0)) or
+                    ((XY[i, dim] > 1.0 - R) and (V[i, dim] > 0))):
+                    V[i, dim] *= -1
+
+        # bounce off of each other
+        #
+        # SUBPROBLEM 1: parallelize this loop over 4 threads, with static
+        # scheduling.
+        for i in prange(count, num_threads=nthreads, chunksize=chunk_size, schedule='static'):
+            sub_update(XY, V, R, i, count, Grid, grid_spacing)
+
+        # update positions
+        #
+        # SUBPROBLEM 1: parallelize this loop over 4 threads (with static
+        #    scheduling).
+        # SUBPROBLEM 2: update the grid values.
+        for i in prange(count, num_threads=nthreads, chunksize=chunk_size, schedule='static'):
+
+            old_grid_x = int(XY[i, 0] / grid_spacing)
+            old_grid_y = int(XY[i, 1] / grid_spacing)
+
+            for dim in range(2):
+                XY[i, dim] += V[i, dim] * t
+
+            # Now update the grid according to the new coordinates
+            # If two fall in the same point, then just overwrite and take the latest ones
+            new_grid_x = int(XY[i, 0] / grid_spacing)
+            new_grid_y = int(XY[i, 1] / grid_spacing)
+
+            # Also set our old point to empty since we aren't there any more
+            # This may overwrite someone who IS there already - e.g., say particle 1 and particle 2
+            # are in the same grid point, but particle 2 was put in there later, so its value is stored in grid.
+            # This shouldn't matter because we are sweeping over the whole grid and updating everything anyways
+            Grid[old_grid_x, old_grid_y] = -1
+            Grid[new_grid_x, new_grid_y] = i
+
+
+def preallocate_locks(num_locks):
+    cdef omp_lock_t *locks = get_N_locks(num_locks)
+    assert 0 != <uintptr_t> <void *> locks, "could not allocate locks"
+    return <uintptr_t> <void *> locks
```

Comment:
Overall, a really nice homework.  You have clearly spent a lot of time thinking about these topics and writing really clean, elegant code.